Split IAM Policy from IAM Role & Improve DependsOn for Streams

alexcasalboni opened the pull request:
## What did you implement:
I have defined a new resource `IamRoleLambdaExecutionPolicy`, as an addition to the `IamRoleLambdaExecution` inlined policy which used to contain all the permissions.
This allows us to define more granular `dependsOn` conditions. For example, we can fix the DDB/Kinesis Streams permission problem in the reference issue.

Closes #3997 

## How did you implement it:

As suggested by @flyinbutrs, I solved the problem by defining a `dependsOn` condition on the IAM Policy. In order to associate a logic name to it, I had to declare a new `AWS::IAM::Policy` and add all the `iamRoleStatements` to it. I couldn't just move all the permissions to the new Policy because a the codebase and many plugins assume that an inlined IAM Policy exists.

## How can we verify it:

You can reproduce the bug with this initial configuration:

```yaml
service: xxx
provider:
  name: aws
  runtime: nodejs10.x
functions:
  myFunction:
    handler: handler.hello
```

Then you deploy it, and update the config to:

```yaml
service: xxx
provider:
  name: aws
  runtime: nodejs6.10
  iamRoleStatements:
    - Effect: "Allow"
      Action:
        - kinesis:GetRecords
        - kinesis:GetShardIterator
        - kinesis:DescribeStream
        - kinesis:ListStreams
      Resource:
        - Fn::GetAtt:
          - MyKinesisStream
          - Arn

functions:
  myFunction:
    handler: handler.hello
    events:
      - stream:
          type: kinesis
          batchSize: 100
          arn:
            Fn::GetAtt:
              - MyKinesisStream
              - Arn

resources:
  Resources:
    MyKinesisStream:
      Type: AWS::Kinesis::Stream
      Properties:
        ShardCount: 1
```

You should get an error similar to the following:

> An error occurred: MyFunctionEventSourceMappingMyKinesisStream - Cannot access stream arn:aws:kinesis:us-east-1:XXX:stream/YYY. Please ensure the role can perform the GetRecords, GetShardIterator, DescribeStream, and ListStreams Actions on your stream in IAM.

The proposed edits fix this problem, as suggested.

## Todos:

- [X] Write tests
- [X] Fix linting errors
- [X] Make sure code coverage hasn't dropped
- [X] Provide verification config / commands / resources
- [X] Enable "Allow edits from maintainers" for this PR
- [X] Update the messages below

***Is this ready for review?:*** NO
***Is it a breaking change?:*** NO

HyperBrain commented:
I think this is a breaking change, as it creates a new policy resource for all existing deployed services?
Wouldn't it be better to opt-in into this mechanism? I fear that plugins that work on the stack (like `serverless-aws-alias` and the like) will be screwed up.

Imo it even would be better if we do not introduce a new fixed policy declaration, but find a solution where the use can define arbitrary policies that can be attached to the lambda role at free will.
I.e. if you'd define IamRolePolicyXXXXXX Serverless could automatically attach these policies to the role.
However, I suggest to keep the default behavior as is.

alexcasalboni commented:
@HyperBrain you are right, it's a breaking change (description updated!).

Ideally, we should have provided plugins with an abstract way of obtaining and modifying the default IAM Policy (something like` provider.getDefaultIAMPolicy()` or `provider.addStatementToDefaultIAMPolicy(...)`), rather than forcing them to inspect and manipulate the raw template directly.

Do you think we could merge this breaking change in two steps? The first one would be adding such a layer and refactoring the plugins you mentioned (is there an easy way to find & notify them?).

An alternative would be leaving the default inlined policy right there and just creating a new independent `AWS::IAM::Policy` that will only contain the list of `iamRoleStatements`. This should not break any plugin (right?) and still provide a way to `dependOn` it.

What do you think?

HyperBrain commented:
> we should have provided plugins with an abstract way of obtaining and modifying the default IAM Policy
That sounds like a good improvement! Regarding the details, I'd rather invent a new `provider.Policy` class that allows full acess to the policy (including add, remove, update, etc). So the Policy class would be something like a datastore, which makes most sense to me. `provider.getDefaultIAMPolicy` should then return the class instance. This would even allow plugins to add policies without knowing any details.
Of course this mechanism can only be used for plugins that act during the package phase. The ones that are triggered during the deployment phase still have to operate on the generated CF template (which is built at the very end of the package phase).

> is there an easy way to find & notify them?
The best way would be to go through the [plugin list](https://github.com/serverless/serverless#v1-plugins) and check if they are affected. However, the most significant issue is the current lack of Serverless to support full semver (major version increase). So the users' projects will unconditionally update to any 1.x SLS version that introduces breaking changes. The plugins cannot specify a meaningful range for the framework version like (>=1.0 <2.0) to prevent them from breaking.

For the `serverless-aws-alias` plugin, I can add a corresponding implementation - but as said before, it would currently not be possible to restrict the correct version without having to know the exact 1.x SLS versions that are compatible or not within the plugin. For the policies there is already a different "workaround" in the plugin, because serverless changed the behavior somewhere between 1.7 and 1.8 as far as I remember. BTW: The alias plugin is a pure "deploy phase" plugin and must operate on the CF template, as it transforms SLS and user resources and just does not care about anything else than the template.
So the only issue here is, to force users to upgrade the plugin as soon as the serverless version containing the change is released (which is nearly impossible because of the 1.x version semantics).

The other plugins have to be checked in detail, if they do any modifications of the policies and rely on the current behavior.

alexcasalboni commented:
@HyperBrain as an intermediate step, what about breaking the assumption that all the IAM statements are part of the same inline Policy? Do you see any breaking change in creating a new independent policy that will contain every statement defined in `iamRoleStatements`?

This way we wouldn't remove direct access to the inline policy and we can move forward with this PR.

HyperBrain commented:
@alexcasalboni Good idea 👍 . Let's go that way.

alexcasalboni commented:
@HyperBrain Done.

I have updated the description and marked this PR as not breaking. Although it may still break a plugin if:

- the plugin assumes that Kinesis/DDB permissions are added to the inline policy
- the plugin assumes that `iamRoleStatements` are added to the inline policy

Imho, such plugins should not rely on such assumption as it's such a core-implementation detail of the framework.

Since the goal of this PR was to fix #3997, I have moved to the new custom IAM Policy only Kinesis and DDB IAM statements.

HyperBrain commented:
> Imho, plugins should never rely on such assumption as it's such a core-implementation detail of the framework.

Agree. Especially plugins working on CF templates they should handle arbitrary combinations (for _aws alias_ it's my fault 😄 and has to be adjusted there).

I'll have a more detailed look at the new changes this weekend.

alexdebrie commented:
I don't have any comments on the actual code changes here, but major thanks to both of you for working on this. 😄 

The problem itself is a particularly confusing one for users. There's a hacky workaround of deploying the updated role first, then deploying right after with the new `stream` event, but it's not intuitive at all. Very nice fix.

Even better, I really like the discussion around how to do this in a non-breaking way and how the Framework should handle permissions in the future. The suggestion of "we should have provided plugins with an abstract way of obtaining and modifying the default IAM Policy" makes a lot of sense and is something I'd like to see in a V2 version of the Framework.

Awesome work 💥 🙌

alexcasalboni commented:
@HyperBrain any update on this?

pmuens commented:
@eahefnawy @dschep @serverless/framework-maintainers can someone else look into this so that we can merge it soon?

It seems like this new approach on how to handler permissions is pretty compelling.

medikoo commented on lib/plugins/aws/package/lib/iam-role-lambda-execution-template.json:
```diff
@@ -42,4 +42,4 @@
     ]
   }
-}
+}
```
This change should be reverted

medikoo commented on lib/plugins/aws/package/lib/iam-policy-lambda-execution-template.json:
```diff
@@ -0,0 +1,13 @@
+      "Ref": "[TO BE REPLACED]"
+    }]
+  }
+}
```
Missing EOL

medikoo commented on lib/plugins/aws/package/lib/saveCompiledTemplate.js:
```diff
@@ -13,8 +13,20 @@ module.exports = {
+        .Statement;
+      // remove custom IAM Policy if no custom statements (empty array would be invalid for CF)
+      if (!customIAMPolicyStatement.length) {
+        delete compiledTemplate.Resources[this.provider.naming.getPolicyLogicalId()];
```
Is this ok to introduce further changes to compiledTemplate here? Maintance-wise it can be confusing.
So far it was purely about storing compiled template into file, now there's also a cleanup action added here.

Maybe we should rather introduce some dedicated _cleanup/optimize step_, prior _save_ that one, and put it there to make it transparent?

pmuens commented:
Thanks again for taking the time to work on this @alexcasalboni 👍 🙏 💯

It seems like this bug just disappeared. Given that I'd propose that we just close this PR and don't introduce such changes which might break something we haven't covered via manual and automated tests.
